fix(eslint-plugin): [no-unnecessary-type-assertion] correct handling of undefined vs. void

Tamashoo opened the pull request:
<!--
👋 Hi, thanks for sending a PR to typescript-eslint! 💖
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR.
-->

## PR Checklist

- [x] Addresses an existing open issue: fixes #11529 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->
Added special handling for type checking of IIFEs in the no-unnecessary-type-assertion rule.
Introduced an isIIFE type guard and a getUncastType helper function to correctly resolve IIFE return types, converting implicit undefined to void for functions without explicit return type annotations.
This improves the accuracy of unnecessary assertion detection for IIFE expressions.

JoshuaKGoldberg reviewed (changes requested):
This is looking good, thanks for the PR! I poked at the rule in a bunch of different ways and couldn't figure out amy way to break the new implementation. Nice work 🙂.

Just requesting changes on more thorough testing. This area of TypeScript (inferred types, overloads/signatures) tends to cause bugs in lint rules a lot.

Tamashoo commented:
@JoshuaKGoldberg 
Thank you for the review!

I’ve added the minimal test cases you suggested and fixed the existing tests.
I’m planning to add an additional test that involves generics before re-requesting review.

If you have any further thoughts on edge cases worth covering, please let me know.
Thank you!

JoshuaKGoldberg reviewed (approved):
🙌 wonderful, thanks!
